apps/speed.c

```diff
@@ -864,27 +865,29 @@ static int CRYPTO_gcm128_aad_loop(void *args)
     return count;
 }
 
-static long save_count = 0;
-static int decrypt = 0;
-static int EVP_Update_loop(void *args)
+static int EVP_Encrypt_loop(void *args)
 {
     loopargs_t *tempargs = *(loopargs_t **)args;
     unsigned char *buf = tempargs->buf;
     EVP_CIPHER_CTX *ctx = tempargs->ctx;
     int outl, count;
-#ifndef SIGALRM
-    int nb_iter = save_count * 4 * lengths[0] / lengths[testnum];
-#endif
-    if (decrypt)
-        for (count = 0; COND(nb_iter); count++)
-            EVP_DecryptUpdate(ctx, buf, &outl, buf, lengths[testnum]);
-    else
-        for (count = 0; COND(nb_iter); count++)
-            EVP_EncryptUpdate(ctx, buf, &outl, buf, lengths[testnum]);
-    if (decrypt)
-        EVP_DecryptFinal_ex(ctx, buf, &outl);
-    else
-        EVP_EncryptFinal_ex(ctx, buf, &outl);
+
+    for (count = 0; COND(c[D_EVP][testnum]); count++)
+        EVP_EncryptUpdate(ctx, buf, &outl, buf, lengths[testnum]);
+    EVP_EncryptFinal_ex(ctx, buf, &outl);
+    return count;
+}
+
+static int EVP_Decrypt_loop(void *args)
+{
+    loopargs_t *tempargs = *(loopargs_t **)args;
+    unsigned char *buf = tempargs->buf;
+    EVP_CIPHER_CTX *ctx = tempargs->ctx;
+    int outl, count;
+
+    for (count = 0; COND(c[D_EVP][testnum]); count++)
+        EVP_DecryptUpdate(ctx, buf, &outl, buf, lengths[testnum]);
```

Comment:
Would it be easier to add the `decrypt` flag to `loopargs_st` instead of splitting `EVP_Update_loop` in two functions that have a lot of duplicated code?

Inside `EVP_Update_loop` we could retrieve the flag and keep the same code used before to run either `EVP_EncryptUpdate` or `EVP_DecryptUpdate`